ul/src/association/mod.rs

```diff
@@ -16,15 +16,599 @@
+
+    /// Retrieve the user variables that were taken from the server.
+    ///
+    /// It usually contains the maximum PDU length,
+    /// the implementation class UID, and the implementation version name.
+    fn user_variables(&self) -> &[UserVariableItem];
+}
+
+mod private {
+    use crate::{Pdu, pdu::{AbortRQServiceProviderReason, AbortRQSource}};
+    use snafu::{ResultExt};
+
+    /// Private trait which exposes "unsafe" methods that should not be called by the user
+    /// 
+    /// `close` and `release` _should_ take ownership, and in the public interface, they 
+    /// do. However, in order to implement `Drop` we need to expose a version of these 
+    /// methods that don't take ownership.
+    /// 
+    /// `send` and `receive` implementations are needed in order to provide
+    /// the implementation for `release`
+    pub trait SyncAssociationSealed<S: std::io::Read + std::io::Write + super::CloseSocket> {
+
+        fn close(&mut self) -> std::io::Result<()>;
+        fn send(&mut self, pdu: &Pdu) -> super::Result<()>;
+        fn receive(&mut self) -> super::Result<Pdu>;
+        fn release(&mut self) -> super::Result<()>{
+            let pdu = Pdu::ReleaseRQ;
+            self.send(&pdu)?;
+            let pdu = self.receive()?;
+
+            match pdu {
+                Pdu::ReleaseRP => {}
+                pdu @ Pdu::AbortRQ { .. }
+                | pdu @ Pdu::AssociationAC { .. }
+                | pdu @ Pdu::AssociationRJ { .. }
+                | pdu @ Pdu::AssociationRQ { .. }
+                | pdu @ Pdu::PData { .. }
+                | pdu @ Pdu::ReleaseRQ => return super::UnexpectedPduSnafu { pdu }.fail(),
+                pdu @ Pdu::Unknown { .. } => return super::UnknownPduSnafu { pdu }.fail(),
+            }
+            self.close()
+                .context(super::CloseSnafu)?;
+            Ok(())
+        }
+
+        fn abort(&mut self) -> super::Result<()> where Self: Sized {
+            let pdu = Pdu::AbortRQ {
+                source: AbortRQSource::ServiceProvider(
+                    AbortRQServiceProviderReason::ReasonNotSpecified,
+                ),
+            };
+            let out = self.send(&pdu);
+            let _ = self.close();
+            out
+        }
+    }
+
+    /// Private trait which exposes "unsafe" methods that should not be called by the user
+    /// 
+    /// `close` and `release` _should_ take ownership, and in the public interface, they 
+    /// do. However, in order to implement `Drop` we need to expose a version of these 
+    /// methods that don't take ownership.
+    /// 
+    /// `send` and `receive` implementations are needed in order to provide
+    /// the implementation for `release`
+    #[cfg(feature = "async")]
+    pub trait AsyncAssociationSealed<S: tokio::io::AsyncRead + tokio::io::AsyncWrite + Unpin> {
+        fn close(&mut self) -> impl std::future::Future<Output = std::io::Result<()>> + Send
```

Comment:
Realizing now thanks to CI that this would require bumping MSRV to 1.75.0

Reply:
Could change to using [async_trait](https://docs.rs/async-trait/latest/async_trait/#explanation) or just changing to `Pin<Box<dyn Future...>>` like `async_trait` does